README.md
# cache-api

Simple application that will purge cache from nginx (using https://github.com/DreamLab/ngx_cache_purge), Cloudflare and 
strapi apollo persistent queries from redis

# Usage

Run server
```bash
$ API_KEY=foqwe NGINX_URL=http://10.8.0.50 CF_API_KEY=Key CF_API_EMAIL=something@at.com CF_ZONE_ID=zoneId go run main.go                                                                                                                     
```

Call api

```bash 
curl -X POST -u api:foqwe  http://172.23.62.218:8080/purge -d '{ "urls": ["https://mkaciuba.pl/graphql?operationName=categoryBySlug&variables=%7B%22categorySlug%22%3A%22milena-studio-2021%22%2C%22limit%22%3A20%2C%22start%22%3A20%7D&extensions=%7B%22persistedQuery%22%3A%7B%22version%22%3A1%2C%22sha256Hash%22%3A%22cfdbf11f396b8da29b65008239b344fc8eda1f74a402e1031b332a9e643f0b95%22%7D%7D"]}'
```



CHANGELOG.md
## [1.0.2](https://github.com/aldor007/purge-api/compare/v1.0.1...v1.0.2) (2022-10-23)


### Bug Fixes

* purge api ([ec98ffd](https://github.com/aldor007/purge-api/commit/ec98ffd450c2158442197f60167d38124c60fe6c))

## [1.0.1](https://github.com/aldor007/purge-api/compare/v1.0.0...v1.0.1) (2022-10-23)


### Bug Fixes

* cf purge ([e412501](https://github.com/aldor007/purge-api/commit/e4125011c067839125705df58828b77339d2bd88))

# 1.0.0 (2022-10-23)


### Bug Fixes

* fix k8s healthcheck ([976b72c](https://github.com/aldor007/purge-api/commit/976b72c986f6442c6ff06a9ebad110e33a4bdf0b))
* fix strapi purge ([1776058](https://github.com/aldor007/purge-api/commit/17760586964e0ec21bf6b5f50facb0727985320c))


### Features

* apollo redis cache purge, dockerfile, chart ([0de324c](https://github.com/aldor007/purge-api/commit/0de324ce40065b1c44ffaaa3f214c2263084f72c))
* new purge and CI ([ff1aa5e](https://github.com/aldor007/purge-api/commit/ff1aa5ef7fb17726a108de14e226ad90328672e1))
* new purge and CI ([2143343](https://github.com/aldor007/purge-api/commit/214334318d9228677652579bb79f81afbc355774))
* support urls ([c6bd228](https://github.com/aldor007/purge-api/commit/c6bd228f983a15e8c22dee4f0bc4c2bb08ee6428))
* update chart ([cda19bd](https://github.com/aldor007/purge-api/commit/cda19bd71df44a5b7ad48f151ae3a9d458b305f7))


cmd/cli.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/fatih/color"
	"github.com/urfave/cli"
)

var user string
var password string

func callAPI(method, url string, body io.Reader) (*http.Response, error) {
	client := &http.Client{
		Transport: &http.Transport{
			Dial: (&net.Dialer{
				Timeout:   30 * time.Second,
				KeepAlive: 3 * time.Second,
			}).Dial,
			TLSHandshakeTimeout:   10 * time.Second,
			ResponseHeaderTimeout: 10 * time.Second,
			ExpectContinueTimeout: 1 * time.Second,
		},
	}
	color.Blue("Calling %s", url)
	req, err := http.NewRequest(method, url, body)
	req.SetBasicAuth(user, password)
	if err != nil {
		return nil, err
	}
	response, err := client.Do(req)
	if err != nil {
		return nil, err
	}


	return response, err
}


func main()  {
	var apiAddress string
	var reqHeaders cli.StringSlice

	app := cli.NewApp()
	app.Name = "purge CLI"
	app.Version = "1.0.0"
	app.Description = "CLI for purge cache and remove media"
	app.UsageText= "CLI for purge cache and remove media"
	app.Flags = []cli.Flag {
		cli.StringFlag{
			Name: "api",
			Value: "http://purge-api.k8s.m39",
			Usage: "API address",
			Destination: &apiAddress,
		},
		cli.StringSliceFlag{
			Name: "header",
			Usage: "Additional requests headers header_name:header_value",
			Value: &reqHeaders,
		},
		cli.StringFlag{
			Name: "user",
			Value: "api",
			Usage: "API address",
			Destination: &user,
		},
		cli.StringFlag{
			Name: "password",
			Value: "123-qwe",
			Usage: "API address",
			Destination: &password,
		},
	}
	app.Commands = []cli.Command{
		{
			Name:    "purge",
			Aliases: []string{"p"},
			Usage:   "purge list of urls",
			Action:  func(c *cli.Context) error {
				data := make(map[string][]string)
				data["urls"] = c.Args()
				body, err := json.Marshal(data)
				if err != nil {
					return err
				}

				headers := make(map[string]string)
				for _, h := range reqHeaders {
					parts := strings.Split(h, ":")
					headers[parts[0]] = parts[1]
				}

				res, err := callAPI("POST", apiAddress + "/purge", bytes.NewReader(body))
				if err != nil {
					return err
				}

				if res.StatusCode == 202 {
					color.Green("Successfully purge %s", data["urls"])
				} else {
					color.Red("There was an error. SC %d", res.StatusCode)
					fmt.Println()
					for h, v := range res.Header {
						fmt.Println("       ", h, ":", v)
					}

					b, _ := ioutil.ReadAll(res.Body)
					fmt.Println(string(b))

				}
				return nil
			},
		},
	}

	err := app.Run(os.Args)
	if err != nil {
		log.Fatal(err)
	}

}


cache/nginx.go
package cache

import (
	"context"
	"errors"
	"fmt"
	"log"
	"net"
	"net/http"
	"net/url"
	"time"
)

type NginxPurgeConfig struct {
	PurgeMethod string
	URL string
}


type Nginx struct {
	client *http.Client
	config NginxPurgeConfig
	endpoint string
}

func NewNginx(config NginxPurgeConfig) *Nginx {

	client := &http.Client{
		Transport: &http.Transport{
			Dial: (&net.Dialer{
				Timeout:   30 * time.Second,
				KeepAlive: 3 * time.Second,
			}).Dial,
			TLSHandshakeTimeout:   10 * time.Second,
			ResponseHeaderTimeout: 10 * time.Second,
			ExpectContinueTimeout: 1 * time.Second,
		},
	}
	u, _  := url.Parse(config.URL)

	return &Nginx{config: config, client: client, endpoint:  u.Host}
}

func (n *Nginx) createPurgeRe(ctx context.Context, url string, headers map[string]string) (*http.Request, error) {
	req, err := http.NewRequestWithContext(ctx, n.config.PurgeMethod, url, nil)
	if err != nil {
		return nil, err
	}

	req.URL.Host = n.endpoint
	req.URL.Scheme = "http"
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	return  req, nil
}
func (n *Nginx) Purge(ctx context.Context, url string) error  {
	purgeReq := make(chan *http.Request, 5)
	errorsChan := make(chan error)
	go func() {
		for req := range purgeReq {
			res, err := n.client.Do(req)
			if err != nil {
				errorsChan <- err
			}

			log.Println("Nginx purge", url, res.StatusCode, req.Header)
			if res.StatusCode != 200 && res.StatusCode != 404 {

				errorsChan <- errors.New(fmt.Sprintf("http error %d", res.StatusCode))
			}
		}
		close(errorsChan)
	}()


	req, err := n.createPurgeRe(ctx, url, map[string]string{"x-forwarded-proto": "https"})
	if err != nil {
		return err
	}
	purgeReq <- req
	req, err = n.createPurgeRe(ctx, url, map[string]string{"x-forwarded-proto": "https", "accept-encoding": "gzip, br"})
	if err != nil {
		return err
	}
	purgeReq <- req
	req, err = n.createPurgeRe(ctx, url, map[string]string{"x-forwarded-proto": "https", "accept-encoding": "gzip"})
	if err != nil {
		return err
	}
	purgeReq <- req
	req, err = n.createPurgeRe(ctx, url, map[string]string{"x-forwarded-proto": "https", "accept-encoding": "gzip", "x-origin-method": "GET"})
	if err != nil {
		return err
	}
	purgeReq <- req
	req, err = n.createPurgeRe(ctx, url, map[string]string{"x-forwarded-proto": "https", "accept-encoding": "br", "x-origin-method": "GET"})
	if err != nil {
		return err
	}
	purgeReq <- req
	close(purgeReq)


	var mainErr error
	for errP := range errorsChan {
		mainErr = errP
	}

	return mainErr
}


cache/cloudflare.go
package cache

import (
	"context"
	"github.com/cloudflare/cloudflare-go"
	"log"
)

type CloudflareConfig struct {
	APIKey string
	APIEmail string
	ZoneID string
}


type Cloudflare struct {
	client *cloudflare.API
	config CloudflareConfig
}

func NewCloudflare(config CloudflareConfig) (*Cloudflare, error) {


	api, err := cloudflare.New(config.APIKey, config.APIEmail)

	if err != nil {
		return nil, err
	}

	return &Cloudflare{config: config, client: api}, nil
}

func (c *Cloudflare) Purge(ctx context.Context,url string) error  {
	pReq := cloudflare.PurgeCacheRequest{
		Everything: true,
		//Files:      []string{url},
		Tags:       nil,
		Hosts:      nil,
		Prefixes:  nil,
	}
	res, err := c.client.PurgeCacheContext(ctx, c.config.ZoneID, pReq)
	if err != nil {
		log.Println("CF purge err", err)
		return err
	}
	if res.Success {
		log.Println("CF purge ", url)
	}
	return nil
}


cache/apollo-strapi.go
package cache

import (
	"context"
	"encoding/json"
	"fmt"
	"github.com/go-redis/redis/v8"
	"net/url"
	"strings"
)


type ApolloStrapiConfig struct {
	RedisEndpoint string
	RedisDB int
}

type apolloExt struct {
	PersistedQuery struct {
		Version    int    `json:"version"`
		Sha256Hash string `json:"sha256Hash"`
	} `json:"persistedQuery"`
}
type ApolloStrapiRedis struct {
	config ApolloStrapiConfig
	client *redis.Client
}

func NewApolloStrapiRedis(config ApolloStrapiConfig) (*ApolloStrapiRedis) {
	client :=  redis.NewClient(&redis.Options{
		Addr:     config.RedisEndpoint ,
		Password: "", // no password set
		DB:       config.RedisDB,  // use default DB
	})

	return &ApolloStrapiRedis{
		client: client,
		config: config,
	}
}

func (a *ApolloStrapiRedis) Purge(ctx context.Context,urlToPurge string) error  {
	if !strings.Contains(urlToPurge, "/graphql") {
		return nil
	}

	decodedValue, err := url.QueryUnescape(urlToPurge)
	if err != nil {
		return err
	}

	p, err := url.Parse(decodedValue)
	if err != nil {
		return err
	}

	exts := p.Query().Get("extensions")
	if exts== "" {
		return nil
	}

	var ext apolloExt
	err = json.Unmarshal([]byte(exts), &ext)
	if err != nil {
		return err
	}
	if ext.PersistedQuery.Sha256Hash == "" {
		return nil
	}
	key := fmt.Sprintf("apq:%s", ext.PersistedQuery.Sha256Hash)
	result := a.client.Del(ctx, key)
	if result.Err() != nil {
		return result.Err()
	}

	return nil
}


cache/mkaciuba-front.go

package cache

import (
	"context"
	"errors"
	"fmt"
	"log"
	"net"
	"net/http"
	"strings"
	"time"
)



type mkaciubaFront struct {
	client *http.Client
	endpoint string
	apiKey string
}

func NewMKaciubaFront(endpoint string, apiKey string) *mkaciubaFront {
	client := &http.Client{
		Transport: &http.Transport{
			Dial: (&net.Dialer{
				Timeout:   30 * time.Second,
				KeepAlive: 3 * time.Second,
			}).Dial,
			TLSHandshakeTimeout:   10 * time.Second,
			ResponseHeaderTimeout: 10 * time.Second,
			ExpectContinueTimeout: 1 * time.Second,
		},
	}

	return &mkaciubaFront{apiKey: apiKey, client: client, endpoint:  endpoint}
}

func (n *mkaciubaFront) createPurgeRe(ctx context.Context, url string, headers map[string]string) (*http.Request, error) {
	toPurge := strings.Replace(url, "https://mkaciuba.pl", "", 1)
	if !strings.HasPrefix(toPurge, "/" ) {
		toPurge = "/" + toPurge
	}
	req, err := http.NewRequestWithContext(ctx, "DELETE", n.endpoint + "?path=" + toPurge , nil)
	if err != nil {
		return nil, err
	}

	for k, v := range headers {
		req.Header.Set(k, v)
	}
	req.Header.Set("x-api-key", n.apiKey)

	return  req, nil
}
func (n *mkaciubaFront) Purge(ctx context.Context, url string) error  {
	if !strings.Contains(url, "mkaciuba.pl") {
		return nil
	}

	purgeReq := make(chan *http.Request, 5)
	errorsChan := make(chan error)
	go func() {
		for req := range purgeReq {
			res, err := n.client.Do(req)
			if err != nil {
				errorsChan <- err
			}

			log.Println("mkaciubaFront purge", url, res.StatusCode, req.Header, req.URL.String())
			if res.StatusCode != 201 && res.StatusCode != 404 {

				errorsChan <- errors.New(fmt.Sprintf("http error %d", res.StatusCode))
			}
		}
		close(errorsChan)
	}()


	req, err := n.createPurgeRe(ctx, url, map[string]string{"x-forwarded-proto": "https"})
	if err != nil {
		return err
	}
	purgeReq <- req
	req, err = n.createPurgeRe(ctx, url, map[string]string{"x-forwarded-proto": "https", "accept-encoding": "gzip, br"})
	if err != nil {
		return err
	}
	purgeReq <- req
	req, err = n.createPurgeRe(ctx, url, map[string]string{"x-forwarded-proto": "https", "accept-encoding": "gzip"})
	if err != nil {
		return err
	}
	purgeReq <- req
	req, err = n.createPurgeRe(ctx, url, map[string]string{"x-forwarded-proto": "https", "accept-encoding": "gzip", "x-origin-method": "GET"})
	if err != nil {
		return err
	}
	purgeReq <- req
	req, err = n.createPurgeRe(ctx, url, map[string]string{"x-forwarded-proto": "https", "accept-encoding": "br", "x-origin-method": "GET"})
	if err != nil {
		return err
	}
	purgeReq <- req
	close(purgeReq)


	var mainErr error
	for errP := range errorsChan {
		mainErr = errP
	}

	return mainErr
}


main.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/aldor007/purge-api/cache"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"strconv"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/go-chi/chi/v5/middleware"
)

func main() {

	purger := cache.NewPurger()

	if os.Getenv("STRAPI_REDIS") != "" {
		redisDB, err :=  strconv.Atoi(os.Getenv("STRAPI_REDIS_DB"))
		if err != nil {
			panic(err)
		}
		purger.AddCache(cache.NewApolloStrapiRedis(cache.ApolloStrapiConfig{
			RedisEndpoint: os.Getenv("STRAPI_REDIS"),
			RedisDB:   redisDB,
		}))
	}

	if os.Getenv("MKACIUBA_FRONT_API") != "" {
		purger.AddCache(cache.NewMKaciubaFront(os.Getenv("MKACIUBA_FRONT_API"), os.Getenv("MKACIUBA_FRONT_API_KEY")))
	}
	if os.Getenv("NGINX_URL") != "" {
		purger.AddCache(cache.NewNginx(cache.NginxPurgeConfig{
			PurgeMethod: "PURGE",
			URL:         os.Getenv("NGINX_URL"),
		}))
	}
	if os.Getenv("CF_API_KEY") != "" {
		cf, err := cache.NewCloudflare(cache.CloudflareConfig{
			APIKey:   os.Getenv("CF_API_KEY"),
			APIEmail: os.Getenv("CF_API_EMAIL"),
			ZoneID:   os.Getenv("CF_ZONE_ID"),
		})
		if err != nil {
			panic(err)
		}
		purger.AddCache(cf)
	}

	r := chi.NewRouter()
	r.Get("/health", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte("welcome"))
	})

	purgeRouter := chi.NewRouter()
	purgeRouter.Use(middleware.Logger)
	purgeRouter.Use(middleware.BasicAuth("purge-api", map[string]string{"api": os.Getenv("API_KEY")}))
	purgeRouter.Get("/", func(w http.ResponseWriter, r *http.Request) {
		toPurge := r.URL.Query().Get("url")
		if toPurge == "" {
			w.WriteHeader(400)
			return
		}

		err := purger.Purge(r.Context(), toPurge)
		if err != nil{
			log.Println("Err purge", toPurge, err)
			w.WriteHeader(500)
			return
		}
		w.WriteHeader(200)
	})
	purgeRouter.Post("/", func(w http.ResponseWriter, r *http.Request) {
		if r.Body == nil {
			w.WriteHeader(400)
			return
		}

		defer r.Body.Close()
		body, err := ioutil.ReadAll(r.Body)
		if err != nil {
			log.Println("Read error", err)
			w.WriteHeader(400)
			return
		}
		purgeData := make(map[string][]string)
		err = json.Unmarshal(body, &purgeData)
		if err != nil {
			log.Println("Parse error", err)
			w.WriteHeader(400)
			return
		}
		toPurgeUrls, ok:= purgeData["urls"]
		if !ok {
			log.Println("No url key")
			w.WriteHeader(400)
			return
		}
		if len(toPurgeUrls) == 0 {
			log.Println("urls empty")
			w.WriteHeader(400)
			return
		}

		errs := make([]error, 0)
		for _, toPurge := range toPurgeUrls {
			err = purger.Purge(r.Context(), toPurge)
			if err != nil{
				log.Println("Err purge", toPurge, err)
				errs = append(errs, err)
			}

		}
		if len(errs) == 0 {
			w.WriteHeader(202)
		} else {
			w.WriteHeader(500)
			fmt.Fprint(w, errs)
		}
	})
	r.Mount("/purge", purgeRouter)
	srv := &http.Server{
		Addr:         ":8080",
		// Good practice to set timeouts to avoid Slowloris attacks.
		WriteTimeout: time.Second * 15,
		ReadTimeout:  time.Second * 15,
		IdleTimeout:  time.Second * 60,
		Handler: r, // Pass our instance of gorilla/mux in.
	}

	log.Println("Server listening :8080")
	if err := srv.ListenAndServe(); err != nil {
		log.Println(err)
	}
}